README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rusleR <img src="man/figures/logo.png" align="right" height="139" />

<!-- badges: start -->

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.6992394.svg)](https://doi.org/10.5281/zenodo.6992394)
[![R-CMD-check](https://github.com/atsyplenkov/rusleR/workflows/R-CMD-check/badge.svg)](https://github.com/atsyplenkov/rusleR/actions)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/rusleR)](https://cran.r-project.org/package=rusleR)
[![Last-changedate](https://img.shields.io/badge/last%20change-2022--12--08-yellowgreen.svg)](/commits/master)
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Project Status: Concept – Minimal or no implementation has been done
yet, or the repository is only intended to be a limited example, demo,
or
proof-of-concept.](https://www.repostatus.org/badges/latest/concept.svg)](https://www.repostatus.org/#concept)

<!-- badges: end -->

This package offers an `R` implementation of Universal Soil Loss
Equation
([USLE](https://en.wikipedia.org/wiki/Universal_Soil_Loss_Equation)).
You can find here a collection of functions to estimate main factors:
R-factor, K-factor, LS-factor and C-factor. The package
uses[`terra`](https://github.com/rspatial/terra) and
[`Rsagacmd`](https://github.com/stevenpawley/Rsagacmd/) in the
background. [`SAGA GIS`](https://sourceforge.net/projects/saga-gis/)
need to be installed on your machine as well.

> **Disclaimer!** This package is under development right now. It was
> tested on Windows 10 only. However, its operability fully depends on
> `terra` and `Rsagacmd` packages. Therefore it is expected that
> `rusleR` should work on Linux-based machines as well.

## Installation

You can install the development version of rusleR from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("atsyplenkov/rusleR")
```

## Example

The brief introduction to the `rusleR` package were made at the
[IAHS-WasWac workshop](megapolis2022.netlify.app) in December 2022. You
can see it on YouTube:

[![MEGAPOLIS2022](https://yt-embed.live/embed?v=B2ian7Gmodc)](https://youtu.be/B2ian7Gmodc?t=5578)

### LS-factor

This is a basic example which shows you how to calculate LS_alpine
([Schmidt et al.,
2019](https://www.sciencedirect.com/science/article/pii/S2215016119300056)):

``` r
library(rusleR)
library(Rsagacmd)
library(terra)

# initiate a saga object
saga <- saga_gis(raster_backend = "terra")

# load DEM
f <- system.file("extdata/dem.tif", package="rusleR")
DEM <- rast(f)

# calculate LS-alpine
ls <- ls_alpine(dem = DEM)
```

As a result of `ls_alpine()` you receive a `SpatRaster` object:

``` r
ls
#> class       : SpatRaster 
#> dimensions  : 78, 54, 1  (nrow, ncol, nlyr)
#> resolution  : 27.30756, 27.30756  (x, y)
#> extent      : 337615.4, 339090, 4814626, 4816756  (xmin, xmax, ymin, ymax)
#> coord. ref. : WGS 84 / UTM zone 38N (EPSG:32638) 
#> source      : memory 
#> name        : LSalpine 
#> min value   :   0.0000 
#> max value   : 660.0054
```

<img src="man/figures/README-pressure-1.png" width="50%" />

### R-factor

With the help of `get_glored()` function you can quickly download and
crop to AOI a Global Rainfall Erosivity Database map
([GLORED](https://esdac.jrc.ec.europa.eu/content/global-rainfall-erosivity)).
It was created by [*Panagos et al.*
(2017)](https://www.nature.com/articles/s41598-017-04282-8) based on
*in-situ* measurements from 3,625 stations. This GLORED was used to
develop a global erosivity map at 30 arc-seconds(\~1 km) based on a
Gaussian Process Regression(GPR).

``` r
library(rusleR)
library(terra)

f <- system.file("extdata/extent.shp", package="rusleR")
v <- vect(f)

r_factor <- get_glored(v)
```

As a result of `get_glored()` you receive a `SpatRaster` object:

``` r
r_factor
#> class       : SpatRaster 
#> dimensions  : 2, 2, 1  (nrow, ncol, nlyr)
#> resolution  : 809.8197, 809.8197  (x, y)
#> extent      : 337544.9, 339164.5, 4814761, 4816381  (xmin, xmax, ymin, ymax)
#> coord. ref. : +proj=utm +zone=38 +datum=WGS84 +units=m +no_defs 
#> source      : memory 
#> name        :      out 
#> min value   : 682.1308 
#> max value   : 820.8951
```

<img src="man/figures/README-r_plot-1.png" width="50%" />

## Citation

While `rusleR` does not redistribute the data or provide it in any way,
we encourage users to cite original papers when using this package. E.g.
to cite Panagos et al. (2017) when using GLORED, Hengl et al. (2017)
when using SoilGrids and Schmidt et al. (2019) when using
LS<sub>alpine</sub>:

> Panagos, Panos, Pasquale Borrelli, Katrin Meusburger, Bofu Yu, Andreas
> Klik, Kyoung Jae Lim, Jae E. Yang, et al. “Global Rainfall Erosivity
> Assessment Based on High-Temporal Resolution Rainfall Records.”
> Scientific Reports 7, no. 1 (June 23, 2017): 4175.
> <https://doi.org/10.1038/s41598-017-04282-8>.

> Hengl, Tomislav, Jorge Mendes de Jesus, Gerard B. M. Heuvelink, Maria
> Ruiperez Gonzalez, Milan Kilibarda, Aleksandar Blagotić, Wei
> Shangguan, et al. “SoilGrids250m: Global Gridded Soil Information
> Based on Machine Learning.” PLOS ONE 12, no. 2 (February 16, 2017):
> e0169748. <https://doi.org/10.1371/journal.pone.0169748>.

> Schmidt, Simon, Simon Tresch, and Katrin Meusburger. “Modification of
> the RUSLE Slope Length and Steepness Factor (LS-Factor) Based on
> Rainfall Experiments at Steep Alpine Grasslands.” MethodsX 6 (2019):
> 219–29. <https://doi.org/10.1016/j.mex.2019.01.004>.

## To-do list

- [x] K-factor functions
- [x] Watem/Sedem functions
- [x] Vignettes
- [ ] tidy-loader disclaimer (see `tidyterra` package)
- [ ] New radar-based R factor
  (<https://esdac.jrc.ec.europa.eu/themes/satellite-based-global-r-factor>)
- [ ] SDR estimation (see [Batista et al.,
  2021](https://www.sciencedirect.com/science/article/pii/S1364815221000049))
- [ ] Connectivity index (see [Borselli et al.,
  2008](https://www.sciencedirect.com/science/article/pii/S0341816208000982))
- [ ] IC by [Cavalli et al.,
  2013](https://www.sciencedirect.com/science/article/pii/S0169555X12002267)
- [ ] SDR by [Vigiak et al.,
  2012](https://www.sciencedirect.com/science/article/pii/S0169555X11004478)
- [ ] C-factor????
- [ ] Should I use WhiteboxGeo tools in a backend instead of SAGA??

### ACKNOWLEDGEMENTS

*This package was developed in accordance to the Development program of
the Interdisciplinary Scientific and Educational School of M.V.
Lomonosov Moscow State University “Future Planet and Global
Environmental Change”*


NEWS.md
# rusleR 0.0.3

* Added WATEM/SEDEM functions

# rusleR 0.0.2

* Added a possibility to choose prediction quantile in `get_soilgrids` function.
* Update to R 4.2.1

---

# rusleR 0.0.1

* Added a `NEWS.md` file to track changes to the package.


R/ls_alpine.R
#' \eqn{LS_{Alpine}}
#'
#' @description This function calculates an LS-factor from Universal Soil
#' Loss Equation (USLE), as described in \emph{Schmidt et al.} (2019).
#'
#' @param dem SpatRaster. A single layer with elevation values. Values should have the same unit as the map units, or in meters when the crs is longitude/latitude
#' @param threshold Numeric. Value specifying a flow threshold in meters. Default is 120 m
#' @param saga_obj SAGA-GIS geoprocessor object. Run \code{Rsagacmd::saga_gis(raster_backend = "terra")}
#'
#' @return SpatRaster
#'
#' @references Schmidt, Simon, Simon Tresch, and Katrin Meusburger. “Modification of the RUSLE Slope Length and Steepness Factor (LS-Factor) Based on Rainfall Experiments at Steep Alpine Grasslands.” MethodsX 6 (2019): 219–29. https://doi.org/10.1016/j.mex.2019.01.004.
#'
#' @examples
#' \dontrun{
#' library(Rsagacmd)
#'
#' # initiate a saga object
#' saga <- saga_gis(raster_backend = "terra")
#'
#' # load DEM
#' f <- system.file("extdata/dem.tif", package="rusleR")
#' DEM <- rast(f)
#'
#' # calculate LS-alpine
#' ls <- ls_alpine(dem = DEM)
#'
#' plot(ls)
#' }
#'
#'
#' @import terra
#' @import Rsagacmd
#'
#' @export
ls_alpine <-
  function(dem,
           threshold = 120,
           saga_obj = saga){

    # Spatial resolution of input DEM
    spatres <- terra::res(dem)[1]
    dem_crs <- terra::crs(dem)

    # Calculate slope in radians
    slope_rad <-
      saga_obj$ta_morphometry$slope_aspect_curvature(
        dem,
        method = "poly2zevenbergen",
        slope = tempfile(fileext = ".sgrd"),
        unit_slope = 0,
        .all_outputs = F
      )

    # Calculate slope in percentage
    slope_perc <-
      saga_obj$ta_morphometry$slope_aspect_curvature(
        dem,
        method = "poly2zevenbergen",
        slope = tempfile(fileext = ".sgrd"),
        unit_slope = 2,
        .all_outputs = F
      )

    # Flow accumulation
    flow_acc <-
      saga$ta_hydrology$flow_accumulation_top_down(
        dem,
        # LINEAR_MIN=500, METHOD=5, CONVERGENCE=1.1
        flow = tempfile(fileext = ".sgrd"),
        method = 5,
        linear_min = 500,
        convergence = 1.1,
        .all_outputs = F
      )

    # Setting a threshold (thresh) of 100m to
    # the flow accumulation assuming a spatial
    # resolution (spatres) of 1m of the DEM.
    # We assume that a total of 100 cells with a
    # width of 1m can accumulate
    threshvalue <- threshold * spatres

    flow_acc_tr1 <-
      terra::ifel(flow_acc == 0,
                  spatres,
                  flow_acc)

    FlowAcc_thresh <-
      terra::ifel(flow_acc_tr1 > threshvalue,
                  threshvalue,
                  flow_acc_tr1)

    # Calculation of L-factor
    Beta <-
      (sin(slope_rad) / 0.0896) / (3 * (sin(slope_rad)^0.8) + 0.56)

    Mvalue <-
      Beta / (1 + Beta)

    Lfactor_w_thresh <-
      ((FlowAcc_thresh + (spatres^2))^(Mvalue + 1) - FlowAcc_thresh^(Mvalue + 1))/(spatres^(Mvalue + 2)*22.13^Mvalue)

    # Calculation of S-factor (alpine)
    Salpine <-
      (1/2000)*(slope_perc^2)+0.1795*(slope_perc)-0.4418

    # Remove negative S-factor
    Salpine_cor <-
      terra::ifel(Salpine < 0, 0, Salpine)

    # LS alpine calculation
    LSalpine <-
      Salpine_cor * Lfactor_w_thresh

    terra::set.names(LSalpine, "LSalpine")
    terra::crs(LSalpine) <- dem_crs

    return(LSalpine)

  }
